src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Sockets
+{
+    internal struct SocketAsyncEngineBackend
+    {
+        private readonly int _epollFd;
+
+        private SocketAsyncEngineBackend(int epollFd)
+        {
+            _epollFd = epollFd;
+        }
+
+        public static SocketAsyncEngineBackend Create()
+        {
+            int epollFd = Interop.libc.epoll_create1(Interop.libc.EPOLL_CLOEXEC);
+            if (epollFd == -1)
+            {
+                // TODO: throw an appropriate exception
+                throw new InternalException();
+            }
+
+            return new SocketAsyncEngineBackend(epollFd);
+        }
+
+        private static SocketAsyncEvents GetSocketAsyncEvents(uint events)
+        {
+            return
+                (((events & Interop.libc.EPOLLIN) != 0) ? SocketAsyncEvents.Read : 0) |
+                (((events & Interop.libc.EPOLLOUT) != 0) ? SocketAsyncEvents.Write : 0) |
+                (((events & Interop.libc.EPOLLRDHUP) != 0) ? SocketAsyncEvents.ReadClose : 0) |
+                (((events & Interop.libc.EPOLLHUP) != 0) ? SocketAsyncEvents.Close : 0) |
+                (((events & Interop.libc.EPOLLERR) != 0) ? SocketAsyncEvents.Error : 0);
+        }
+
+        private static uint GetEPollEvents(SocketAsyncEvents events)
+        {
+            return
+                (((events & SocketAsyncEvents.Read) != 0) ? Interop.libc.EPOLLIN : 0) |
+                (((events & SocketAsyncEvents.Write) != 0) ? Interop.libc.EPOLLOUT : 0) |
+                (((events & SocketAsyncEvents.ReadClose) != 0) ? Interop.libc.EPOLLRDHUP : 0) |
+                (((events & SocketAsyncEvents.Close) != 0) ? Interop.libc.EPOLLHUP : 0) |
+                (((events & SocketAsyncEvents.Error) != 0) ? Interop.libc.EPOLLERR : 0);
+        }
+
+        public unsafe void EventLoop()
+        {
+            var events = stackalloc Interop.libc.epoll_event[64];
+            for (;;)
+            {
+                int numEvents = Interop.libc.epoll_wait(_epollFd, events, 64, -1);
```

Comment:
Consider moving the ```64``` into a named constant that can be used both here and in the stackalloc.